### PR TITLE
stub+ukify: Add support for UKI .ucode section

### PR DESCRIPTION
**Pull request by tfg13**
Hello!

I am picking up [this](https://github.com/systemd/systemd/blob/main/TODO#L518) item from the TODO list, and add support for a `.ucode` section in UKIs. This is meant to be the first step, the second PR will be adding support for ucode _addon_ files.

This PR:
  - Declares a new UKI section `.ucode`, which is functionally an initrd, but will always be the first one handed to the kernel (note it can't be named `.microcode`, due to the 8 char limitation in PE binaries)
  - Adds support in ukify to create UKIs with this new section
  - Adds support in sd-stub to load the new section, measure it, and hand it to the kernel as the first initrd

Tested by hand, and tested secureboot by teaching mkosi to use the new flag in this PR: https://github.com/systemd/mkosi/pull/2556

I have a few open questions/comments I'd appreciate some input on:
  - This is effectively treated as an initrd for measurement purposes. I think this makes sense, since there is ultimately no guarantee the archive actually contains (just) microcode. Is that ok?
  - ~Any other place I should add/document the new section name? UKI specs/Docs/manpages?~
  - I know there is also #28070, which does related things, but hasn't been updated in a bit. Happy to chat so we don't step on each others toes

### CODE REVIEW

File: src/boot/efi/stub.c [hunk 1]
```diff
@@ -26,19 +26,30 @@ DECLARE_NOALLOC_SECTION(".sdmagic", "#### LoaderInfo: systemd-stub " GIT_VERSION
 
 DECLARE_SBAT(SBAT_STUB_SECTION_TEXT);
 
-static EFI_STATUS combine_initrd(
-                EFI_PHYSICAL_ADDRESS initrd_base, size_t initrd_size,
+/* Combine initrds by concatenation in memory.
+ * uki_initrds come from sections in the UKI and will be padded,
+ * extra initrds are generated and don't need further padding */
+static EFI_STATUS combine_initrds(
+                EFI_PHYSICAL_ADDRESS uki_initrds[], const size_t uki_initrd_sizes[], size_t n_uki_initrds,
```
Comment (daandemeyer) [inline src/boot/efi/stub.c]: I wouldn't go too fancy here and just add two new arguments to specify microcode base and size and use `0` for no microcode initrd specified.

Comment (daandemeyer) [reply src/boot/efi/stub.c]: You can still treat both the microcode and the actual initrd as an array inside the function though.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]

File: src/boot/efi/stub.c (hunk 1 above)
Comment (tfg13) [reply src/boot/efi/stub.c]: Sure, can do. I figured this makes it a bit more mergeable if #28070 (multiple initrds) want to move forward at some point, but I can keep it simple for now

File: src/boot/efi/stub.c [hunk 2]
```diff
@@ -26,19 +26,30 @@ DECLARE_NOALLOC_SECTION(".sdmagic", "#### LoaderInfo: systemd-stub " GIT_VERSION
...
         assert(ret_initr_pages);
         assert(ret_initrd_size);
 
-        /* Combines four initrds into one, by simple concatenation in memory */
+        for (size_t i = 0; i < n_uki_initrds; i++) {
+                if (!uki_initrds[i])
+                        continue;
 
-        n = ALIGN4(initrd_size); /* main initrd might not be padded yet */
+                // initrds from UKI sections need padding
```
Comment (bluca) [inline src/boot/efi/stub.c]: `/* */` for comments

File: src/ukify/ukify.py
Comment (bluca) [inline src/ukify/ukify.py:1]: please split the ukify change in a separate commit in the same pr

Comment (bluca) [reply src/ukify/ukify.py:1]: this also needs to be documented in the manpage

File: src/boot/measure.c
Comment (bluca) [inline src/boot/measure.c:1]: split in a separate commit, document in manpage

File: src/boot/efi/stub.c
Comment (bluca) [inline src/boot/efi/stub.c:1]: needs to be documented in stub manpage

File: src/boot/efi/stub.c (hunk 1 above)
Comment (bluca) [reply src/boot/efi/stub.c]: given we'll have multiple initrd it's fine with this approach with an array

**Review by bluca — CHANGES_REQUESTED** [review]

**Comment by tfg13** [comment]
Thanks for the reviews! I split the PR into three commits, documented the changes in the manpages, and addressed the other nits.

File: man/systemd-stub.xml [hunk 3]
```diff
@@ -70,6 +70,9 @@
 
       <listitem><para>An <literal>.initrd</literal> section with the initrd.</para></listitem>
 
+      <listitem><para>A <literal>.ucode</literal> section with a initrd containing microcode, to be handed
```
Comment (daandemeyer) [inline man/systemd-stub.xml]: "an initrd"?

I would also explicitly mention that this initrd cannot be compressed.

Comment (tfg13) [reply man/systemd-stub.xml]: Done (both). Also rebased.

**Comment by poettering** [comment]
So the concept definitely looks great to me. But before we merge this I'd like to see some discussion how we intended to handle multiple .ucode sections, see my points here:

https://github.com/uapi-group/specifications/pull/98#issuecomment-2056783969

if we opt for the 3rd option this has repercussions for the .ucode data format. 

It's totally fine if we don#t actually implement support for multiple .ucode sections in add-ons/UKIs. But I'd prefer if we'd at least have an idea what to do in that case...

File: src/boot/efi/stub.c [hunk 4]
```diff
@@ -26,19 +26,30 @@ DECLARE_NOALLOC_SECTION(".sdmagic", "#### LoaderInfo: systemd-stub " GIT_VERSION
...
 DECLARE_SBAT(SBAT_STUB_SECTION_TEXT);
 
-static EFI_STATUS combine_initrd(
-                EFI_PHYSICAL_ADDRESS initrd_base, size_t initrd_size,
+/* Combine initrds by concatenation in memory.
+ * uki_initrds come from sections in the UKI and will be padded,
+ * extra initrds are generated and don't need further padding */
+static EFI_STATUS combine_initrds(
+                EFI_PHYSICAL_ADDRESS uki_initrds[], const size_t uki_initrd_sizes[], size_t n_uki_initrds,
                 const void * const extra_initrds[], const size_t extra_initrd_sizes[], size_t n_extra_initrds,
```
Comment (poettering) [inline src/boot/efi/stub.c]: please just make this one set of arrays. and apply the padding always. if the initrds happen to be padded already it's a NOP to pad them again after all.

File: src/boot/efi/stub.c [hunk 5]
```diff
@@ -806,14 +805,16 @@ static EFI_STATUS run(EFI_HANDLE image) {
...
+                                        initrd_size,
                                         credential_initrd_size,
                                         global_credential_initrd_size,
                                         sysext_initrd_size,
                                         confext_initrd_size,
                                         pcrsig_initrd_size,
                                         pcrpkey_initrd_size,
                                 },
-                                6,
+                                2,
```
Comment (keszybz) [inline src/boot/efi/stub.c]: Hmmm, why "2" here?

File: man/ukify.xml [hunk 6]
```diff
@@ -67,6 +67,7 @@
 
       <para>Additional sections will be inserted into the UKI, either automatically or only if a specific
       option is provided. See the discussions of
+      <varname>Ucode=</varname>/<option>--ucode=</option>,
```
Comment (keszybz) [inline man/ukify.xml]: In the commit message: "functionally"

I think this should be called "Microcode=". The other options are not abbreviated and do not correspond 1:1 to the section names.

File: man/ukify.xml [hunk 7]
```diff
@@ -302,6 +303,16 @@
           <xi:include href="version-info.xml" xpointer="v254"/></listitem>
         </varlistentry>
 
+        <varlistentry>
+          <term><varname>Ucode=<replaceable>UCODE</replaceable></varname></term>
+          <term><option>--ucode=<replaceable>UCODE</replaceable></option></term>
```
Comment (keszybz) [inline man/ukify.xml]: And "--microcode" here.

**Review by keszybz — CHANGES_REQUESTED** [review]

File: src/boot/efi/stub.c (hunk 5 above)
Comment (tfg13) [reply src/boot/efi/stub.c]: ooh merge problem, thanks for catching it. Fixing...

File: man/ukify.xml (hunk 6 above)
Comment (tfg13) [reply man/ukify.xml]: I tried to keep the terminology the same everywhere, and the section has that name. But now I see for the other sections we also use longer, more descriptive names, so I'll change it

**Comment by tfg13** [comment]
One sec I messed up the commits

**Review by keszybz — APPROVED** [review]
LGTM.
